gapis/perfetto/android/trace.go

```diff
@@ -76,26 +91,39 @@ func setupRenderStagesEnvironment(ctx context.Context, d adb.Device, packageName
 	if res, err := d.Shell("pm", "path", driverPackageName).Call(ctx); err != nil || res == "" {
 		return nil, log.Err(ctx, err, "No driver package found.")
 	}
-	cleanup, err := android.SetupLayer(ctx, d, packageName, driverPackageName, renderStageVulkanLayerName, true)
+	packages := []string{driverPackageName}
+	enabledLayers := []string{renderStageVulkanLayerName}
+	if abi != nil {
+		packages = append(packages, gapidapk.PackageName(abi))
+		enabledLayers = append(enabledLayers, layers...)
+	}
+
+	cleanup, err := android.SetupLayers(ctx, d, packageName, packages, enabledLayers, true)
 	if err != nil {
 		return cleanup.Invoke(ctx), log.Err(ctx, err, "Failed to setup gpu.renderstages environment.")
 	}
 	return cleanup, nil
 }
 
 // Start optional starts an app and sets up a Perfetto trace
-func Start(ctx context.Context, d adb.Device, a *android.ActivityAction, opts *service.TraceOptions) (*Process, app.Cleanup, error) {
+func Start(ctx context.Context, d adb.Device, a *android.ActivityAction, opts *service.TraceOptions, abi *device.ABI, layers []string) (*Process, app.Cleanup, error) {
 	ctx = log.Enter(ctx, "start")
 
 	var cleanup app.Cleanup
+	if abi != nil {
+		_, err := gapidapk.EnsureInstalled(ctx, d, abi)
+		if err != nil {
+			return nil, cleanup, err
+		}
+	}
 	if a != nil {
 		ctx = log.V{
 			"package":  a.Package.Name,
 			"activity": a.Activity,
 		}.Bind(ctx)
 		// Before we start the activity, attempt to setup environment if gpu.renderstages is selected.
 		var err error
-		cleanup, err = setupRenderStagesEnvironment(ctx, d, a.Package.Name, opts.PerfettoConfig)
+		cleanup, err := setupRenderStagesEnvironment(ctx, d, a.Package.Name, opts.PerfettoConfig, abi, layers)
```

Comment:
remove the `:`. With it, go creates the new `cleanup` var inside this scope and the outside one stays nil.